### PR TITLE
Enforce read-only semesters

### LINKED ISSUE #21: Make previous semesters as read only
- Disallow edits to previous semesters
- Drop `semester` parameter in PUT / POST requests
  - Add it in the API layer

### PR DESCRIPTION
**Pull request by vikaspotluri123**
closes #21

### CODE REVIEW

**Comment by joshcosta** [comment]
All creates need to be permissified as well

**Comment by vikaspotluri123** [comment]
I ended up going down the rabbit hole of query optimizations and realized core-data can be improved. This was noticed locally (10.4.12-MariaDB-1:10.4.12+maria~bionic mariadb.org binary distribution) but not reproduced in prod (10.3.22-MariaDB-0+deb10u1-log Debian 10). While I'm locally on a newer version, I'm pretty sure prod is just using a different optimization technique due to the sheer amount of data. Either way, I'd like to search by course because there's a much lower limit (MAX_COURSES_PER_SEMESTER) compared to categories (MAX_COURSES_PER_SEMESTER * MAX_CATEGORIES_PER_COURSE).

local: 
```sql
MariaDB [populated]> EXPLAIN SELECT * FROM grades WHERE course_id IN (...5 courses);
+------+-------------+--------+-------+--------------------------+--------------------------+---------+------+------+-----------------------+
| id   | select_type | table  | type  | possible_keys            | key                      | key_len | ref  | rows | Extra                 |
+------+-------------+--------+-------+--------------------------+--------------------------+---------+------+------+-----------------------+
|    1 | SIMPLE      | grades | range | grades_course_id_foreign | grades_course_id_foreign | 98      | NULL | 5    | Using index condition |
+------+-------------+--------+-------+--------------------------+--------------------------+---------+------+------+-----------------------+
1 row in set (0.01 sec)

MariaDB [populated]> EXPLAIN SELECT *  FROM grades WHERE category_id IN (...33 categories);
+------+-------------+--------+------+----------------------------------------+------+---------+------+------+-------------+
| id   | select_type | table  | type | possible_keys                          | key  | key_len | ref  | rows | Extra       |
+------+-------------+--------+------+----------------------------------------+------+---------+------+------+-------------+
|    1 | SIMPLE      | grades | ALL  | grades_category_id_foreign,category_id | NULL | NULL    | NULL | 137  | Using where |
+------+-------------+--------+------+----------------------------------------+------+---------+------+------+-------------+
1 row in set (0.00 sec)
```

Prod:
```sql
MariaDB [aggie]> EXPLAIN SELECT * FROM grades WHERE course_id IN (...5 courses);
+------+-------------+--------+-------+--------------------------+--------------------------+---------+------+------+-----------------------+
| id   | select_type | table  | type  | possible_keys            | key                      | key_len | ref  | rows | Extra                 |
+------+-------------+--------+-------+--------------------------+--------------------------+---------+------+------+-----------------------+
|    1 | SIMPLE      | grades | range | grades_course_id_foreign | grades_course_id_foreign | 98      | NULL |  100 | Using index condition |
+------+-------------+--------+-------+--------------------------+--------------------------+---------+------+------+-----------------------+
1 row in set (0.000 sec)

MariaDB [aggie]> EXPLAIN SELECT *  FROM grades WHERE category_id IN (...33 categories);
+------+-------------+--------+-------+----------------------------+----------------------------+---------+------+------+-----------------------+
| id   | select_type | table  | type  | possible_keys              | key                        | key_len | ref  | rows | Extra                 |
+------+-------------+--------+-------+----------------------------+----------------------------+---------+------+------+-----------------------+
|    1 | SIMPLE      | grades | range | grades_category_id_foreign | grades_category_id_foreign | 98      | NULL |  100 | Using index condition |
+------+-------------+--------+-------+----------------------------+----------------------------+---------+------+------+-----------------------+
1 row in set (0.001 sec)
```

**Comment by vikaspotluri123** [comment]
@joshcosta it's ready now 😄

File: lib/services/current-semester.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+	activeSemester: '2020U',
+	validSemesters: new Set(['2020U']),
+	validSemestersArray: ['2020U']
```
Comment (joshcosta) [inline lib/services/current-semester.js:4]: Why do we need both?

Comment (vikaspotluri123) [reply lib/services/current-semester.js:4]: Sets are used for searching for an item, arrays are used for providing a list of items

**Comment by vikaspotluri123** [comment]
@ramsays merging now but please review the changes and note your thoughts

**Pull request merged**